sklearn/decomposition/nmf.py

```diff
@@ -254,22 +344,287 @@ def _nls_subproblem(V, W, H, tol, max_iter, sigma=0.01, beta=0.1):
+            H, tolH, nls_max_iter, regul=alpha, l1_ratio=l1_ratio)
+    elif sparseness == 'components':
+        H, gradH, iterH = _nls_subproblem(
+            safe_vstack([X, np.zeros((1, n_features))]),
+            safe_vstack([W,
+                         np.sqrt(beta)
+                         * np.ones((1, n_components_))]),
+            H, tolH, nls_max_iter, regul=alpha, l1_ratio=l1_ratio)
+
+    return H, gradH, iterH
+
+
+def _fit_projected_gradient(X, W, H, tol, max_iter,
+                            nls_max_iter, alpha, l1_ratio,
+                            sparseness, beta, eta):
+    """Compute Non-negative Matrix Factorization (NMF) with Projected Gradient
 
-    Read more in the :ref:`User Guide <NMF>`.
+    References
+    ----------
+    C.-J. Lin. Projected gradient methods for non-negative matrix
+    factorization. Neural Computation, 19(2007), 2756-2779.
+    http://www.csie.ntu.edu.tw/~cjlin/nmf/
+
+    P. Hoyer. Non-negative Matrix Factorization with Sparseness Constraints.
+    Journal of Machine Learning Research 2004.
+    """
+    gradW = (np.dot(W, np.dot(H, H.T))
+             - safe_sparse_dot(X, H.T, dense_output=True))
+    gradH = (np.dot(np.dot(W.T, W), H)
+             - safe_sparse_dot(W.T, X, dense_output=True))
+
+    init_grad = squared_norm(gradW) + squared_norm(gradH.T)
+    # max(0.001, tol) to force alternating minimizations of W and H
+    tolW = max(0.001, tol) * np.sqrt(init_grad)
+    tolH = tolW
+
+    for n_iter in range(1, max_iter + 1):
+        # stopping condition
+        # as discussed in paper
+        proj_grad_W = squared_norm(gradW * np.logical_or(gradW < 0, W > 0))
+        proj_grad_H = squared_norm(gradH * np.logical_or(gradH < 0, H > 0))
+
+        if (proj_grad_W + proj_grad_H) / init_grad < tol ** 2:
+            break
+
+        # update W
+        W, gradW, iterW = _update_projected_gradient_w(X, W, H, tolW,
+                                                       nls_max_iter,
+                                                       alpha, l1_ratio,
+                                                       sparseness, beta, eta)
+        if iterW == 1:
+            tolW = 0.1 * tolW
+
+        # update H
+        H, gradH, iterH = _update_projected_gradient_h(X, W, H, tolH,
+                                                       nls_max_iter,
+                                                       alpha, l1_ratio,
+                                                       sparseness, beta, eta)
+        if iterH == 1:
+            tolH = 0.1 * tolH
+
+    H[H == 0] = 0   # fix up negative zeros
+
+    if n_iter == max_iter:
+        W, _, _ = _update_projected_gradient_w(X, W, H, tol, nls_max_iter,
+                                               alpha, l1_ratio, sparseness,
+                                               beta, eta)
+
+    return W, H, n_iter
+
+
+def _update_coordinate_descent(X, W, Ht, alpha, l1_ratio, shuffle,
```

Comment:
I'm not entirely sure about the naming, but maybe this way is easiest? I mean X can be X or X.T and W can be W or H .... but I don't have a better idea ;)